### PR TITLE
During lasso select, mask values using Dask-compatible method

### LINKED ISSUE #5567: Dask DataFrame + Lasso select linked brushing not working
IMPORTANT: I have narrowed down the problem to the following line of code. Assignment like this is not supported by Dask. I will submit an accompanying pull request with a proposed fix.
https://github.com/holoviz/holoviews/blob/f7017e3012ef60d7368676d1c87efe8194d03a1e/holoviews/element/selection.py#L120 

#### ALL software version info
python=3.10.8
bokeh=2.4.3
dask=2022.12.0
datashader=0.14.3
holoviews=1.15.3
numpy=1.23.5
pandas=1.5.2
panel=0.14.1
spatialpandas=0.4.5

#### Description of expected behavior and the observed behavior
Attempting to lasso-select points that are backed by a Dask DataFrame results in an error. Box select works fine, but lasso select does not. This issue exists for both datashaded and non-datshaded plots. Same result using spatialpandas and shapely.

#### Complete, minimal, self-contained example code that reproduces the issue

```python
import dask.dataframe as dd
import holoviews as hv
import panel as pn
from holoviews.operation.datashader import datashade
from numpy.random import randn


hv.extension("bokeh")

df = dd.from_dict({"x": randn(1000), "y": randn(1000)}, npartitions=2)

link = hv.link_selections.instance()

non_datashaded_plot = link(hv.Points(df))
datashaded_plot = link(datashade(hv.Points(df)))
pn.Column(non_datashaded_plot + datashaded_plot).show()

# Error will occur when you try to lasso-select data in either plot
```

#### Stack traceback and/or browser JavaScript console output
```shell
WARNING:param.dynamic_operation: Callable raised "CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')".
Invoked as dynamic_operation(cmap=None, exprs=[True, dim('x', spatial_select, dim('y'), geometry=array([[ 1.81901611,  0.02067943],
       [ 1.9212558 ,  0.3637043 ],
       [ 1.853096  ,  0.70672917],
       [ 1.27373776,  0.73791325],
       [-0.70289623,  0.64436101],
       [-2.23649157,  0.33252022],
       [-2.88400961, -0.04168873],
       [-3.25888847, -0.66537031],
       [-3.49744774, -1.66326085],
       [-3.39520805, -2.50523099],
       [-2.9862493 , -3.12891257],
       [-2.54321064, -3.5343056 ],
       [-1.92977251, -3.690226  ],
       [-1.45265396, -3.72141008],
       [-1.14593489, -3.72141008],
       [-1.0096153 , -3.690226  ],
       [-0.80513592, -3.65904192],
       [-0.77105603, -3.62785784],
       [-0.77105603, -3.62785784]]))])
WARNING:param.dynamic_operation: Callable raised "CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')".
Invoked as dynamic_operation(alpha=0.1, colors=[None, None], cmap=None)
WARNING:param.dynamic_operation: Callable raised "CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')".
Invoked as dynamic_operation(cmap=None, exprs=[True, dim('x', spatial_select, dim('y'), geometry=array([[ 1.81901611,  0.02067943],
       [ 1.9212558 ,  0.3637043 ],
       [ 1.853096  ,  0.70672917],
       [ 1.27373776,  0.73791325],
       [-0.70289623,  0.64436101],
       [-2.23649157,  0.33252022],
       [-2.88400961, -0.04168873],
       [-3.25888847, -0.66537031],
       [-3.49744774, -1.66326085],
       [-3.39520805, -2.50523099],
       [-2.9862493 , -3.12891257],
       [-2.54321064, -3.5343056 ],
       [-1.92977251, -3.690226  ],
       [-1.45265396, -3.72141008],
       [-1.14593489, -3.72141008],
       [-1.0096153 , -3.690226  ],
       [-0.80513592, -3.65904192],
       [-0.77105603, -3.62785784],
       [-0.77105603, -3.62785784]]))])
WARNING:param.dynamic_operation: Callable raised "CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')".
Invoked as dynamic_operation(alpha=0.1, colors=[None, None], cmap=None)
WARNING:param.dynamic_operation: Callable raised "CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')".
Invoked as dynamic_operation(cmap=None, exprs=[True, dim('x', spatial_select, dim('y'), geometry=array([[ 1.81901611,  0.02067943],
       [ 1.9212558 ,  0.3637043 ],
       [ 1.853096  ,  0.70672917],
       [ 1.27373776,  0.73791325],
       [-0.70289623,  0.64436101],
       [-2.23649157,  0.33252022],
       [-2.88400961, -0.04168873],
       [-3.25888847, -0.66537031],
       [-3.49744774, -1.66326085],
       [-3.39520805, -2.50523099],
       [-2.9862493 , -3.12891257],
       [-2.54321064, -3.5343056 ],
       [-1.92977251, -3.690226  ],
       [-1.45265396, -3.72141008],
       [-1.14593489, -3.72141008],
       [-1.0096153 , -3.690226  ],
       [-0.80513592, -3.65904192],
       [-0.77105603, -3.62785784],
       [-0.77105603, -3.62785784]]))])
WARNING:param.dynamic_operation: Callable raised "CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')".
Invoked as dynamic_operation(alpha=0.1, colors=[None, None], cmap=None)
Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fb2d03e7610>>, <Task finished name='Task-1101' coro=<ServerSession.with_document_locked() done, defined at /user_environment/lib/python3.10/site-packages/bokeh/server/session.py:78> exception=CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')>)
Traceback (most recent call last):
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 511, in _select
    mask = selection_expr.apply(dataset, compute=False, keep_index=True, strict=False)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/transform.py", line 773, in apply
    data = self._apply_fn(dataset, data, fn, fn_name, args,
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/transform.py", line 673, in _apply_fn
    data = fn(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/element/selection.py", line 128, in spatial_select
    return spatial_select_columnar(xvals, yvals, geometry)
  File "/user_environment/lib/python3.10/site-packages/holoviews/element/selection.py", line 120, in spatial_select_columnar
    mask[np.where(mask)[0]] = geom_mask
TypeError: only integer scalar arrays can be converted to a scalar index

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/user_environment/lib/python3.10/site-packages/tornado/ioloop.py", line 740, in _run_callback
    ret = callback()
  File "/user_environment/lib/python3.10/site-packages/tornado/ioloop.py", line 764, in _discard_future_result
    future.result()
  File "/user_environment/lib/python3.10/site-packages/bokeh/server/session.py", line 95, in _needs_document_lock_wrapper
    result = func(self, *args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/server/session.py", line 229, in with_document_locked
    return func(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 450, in wrapper
    return invoke_with_curdoc(doc, invoke)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 408, in invoke_with_curdoc
    return f()
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 449, in invoke
    return f(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 179, in remove_then_invoke
    return callback()
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 243, in refresh
    raise e
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 239, in refresh
    self._trigger_refresh(stream_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 256, in _trigger_refresh
    self.update(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 949, in update
    item = self.__getitem__(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 435, in __getitem__
    self.update_frame(frame)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/bokeh/element.py", line 2410, in update_frame
    element = self._get_frame(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 1259, in _get_frame
    frame = get_plot_frame(self.hmap, key_map, cached)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/util.py", line 284, in get_plot_frame
    raise e
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/util.py", line 280, in get_plot_frame
    return map_obj[key]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 550, in __call__
    if not args and not kwargs and not any(kwarg_hash): return self.callable()
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 199, in dynamic_mul
    self_el = self.select(HoloMap, **key_map) if self.kdims else self[()]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 550, in __call__
    if not args and not kwargs and not any(kwarg_hash): return self.callable()
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 204, in dynamic_mul
    other_el = other.select(HoloMap, **key_map) if other.kdims else other[()]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 581, in __call__
    ret = self.callable(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1012, in dynamic_operation
    key, obj = resolve(key, kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1001, in resolve
    return key, map_obj[key]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 581, in __call__
    ret = self.callable(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1013, in dynamic_operation
    return apply(obj, *key, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1005, in apply
    processed = self._process(element, key, kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 989, in _process
    return self.p.operation(element, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 612, in _build_layer_callback
    selection = self._select(element, exprs[layer_number], cache)
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 519, in _select
    raise CallbackError("linked_selection aborted because it could not "
holoviews.core.options.CallbackError: linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.
Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fb2d03e7610>>, <Task finished name='Task-1102' coro=<ServerSession.with_document_locked() done, defined at /user_environment/lib/python3.10/site-packages/bokeh/server/session.py:78> exception=CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')>)
Traceback (most recent call last):
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 511, in _select
    mask = selection_expr.apply(dataset, compute=False, keep_index=True, strict=False)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/transform.py", line 773, in apply
    data = self._apply_fn(dataset, data, fn, fn_name, args,
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/transform.py", line 673, in _apply_fn
    data = fn(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/element/selection.py", line 128, in spatial_select
    return spatial_select_columnar(xvals, yvals, geometry)
  File "/user_environment/lib/python3.10/site-packages/holoviews/element/selection.py", line 120, in spatial_select_columnar
    mask[np.where(mask)[0]] = geom_mask
TypeError: only integer scalar arrays can be converted to a scalar index

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/user_environment/lib/python3.10/site-packages/tornado/ioloop.py", line 740, in _run_callback
    ret = callback()
  File "/user_environment/lib/python3.10/site-packages/tornado/ioloop.py", line 764, in _discard_future_result
    future.result()
  File "/user_environment/lib/python3.10/site-packages/bokeh/server/session.py", line 95, in _needs_document_lock_wrapper
    result = func(self, *args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/server/session.py", line 229, in with_document_locked
    return func(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 450, in wrapper
    return invoke_with_curdoc(doc, invoke)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 408, in invoke_with_curdoc
    return f()
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 449, in invoke
    return f(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 179, in remove_then_invoke
    return callback()
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 243, in refresh
    raise e
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 239, in refresh
    self._trigger_refresh(stream_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 256, in _trigger_refresh
    self.update(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 949, in update
    item = self.__getitem__(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 435, in __getitem__
    self.update_frame(frame)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/bokeh/element.py", line 2410, in update_frame
    element = self._get_frame(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 1259, in _get_frame
    frame = get_plot_frame(self.hmap, key_map, cached)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/util.py", line 284, in get_plot_frame
    raise e
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/util.py", line 280, in get_plot_frame
    return map_obj[key]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 550, in __call__
    if not args and not kwargs and not any(kwarg_hash): return self.callable()
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 199, in dynamic_mul
    self_el = self.select(HoloMap, **key_map) if self.kdims else self[()]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 550, in __call__
    if not args and not kwargs and not any(kwarg_hash): return self.callable()
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 204, in dynamic_mul
    other_el = other.select(HoloMap, **key_map) if other.kdims else other[()]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 581, in __call__
    ret = self.callable(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1012, in dynamic_operation
    key, obj = resolve(key, kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1001, in resolve
    return key, map_obj[key]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 581, in __call__
    ret = self.callable(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1013, in dynamic_operation
    return apply(obj, *key, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1005, in apply
    processed = self._process(element, key, kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 989, in _process
    return self.p.operation(element, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 612, in _build_layer_callback
    selection = self._select(element, exprs[layer_number], cache)
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 519, in _select
    raise CallbackError("linked_selection aborted because it could not "
holoviews.core.options.CallbackError: linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.
Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fb2d03e7610>>, <Task finished name='Task-1104' coro=<ServerSession.with_document_locked() done, defined at /user_environment/lib/python3.10/site-packages/bokeh/server/session.py:78> exception=CallbackError('linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.')>)
Traceback (most recent call last):
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 511, in _select
    mask = selection_expr.apply(dataset, compute=False, keep_index=True, strict=False)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/transform.py", line 773, in apply
    data = self._apply_fn(dataset, data, fn, fn_name, args,
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/transform.py", line 673, in _apply_fn
    data = fn(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/element/selection.py", line 128, in spatial_select
    return spatial_select_columnar(xvals, yvals, geometry)
  File "/user_environment/lib/python3.10/site-packages/holoviews/element/selection.py", line 120, in spatial_select_columnar
    mask[np.where(mask)[0]] = geom_mask
TypeError: only integer scalar arrays can be converted to a scalar index

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/user_environment/lib/python3.10/site-packages/tornado/ioloop.py", line 740, in _run_callback
    ret = callback()
  File "/user_environment/lib/python3.10/site-packages/tornado/ioloop.py", line 764, in _discard_future_result
    future.result()
  File "/user_environment/lib/python3.10/site-packages/bokeh/server/session.py", line 95, in _needs_document_lock_wrapper
    result = func(self, *args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/server/session.py", line 229, in with_document_locked
    return func(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 450, in wrapper
    return invoke_with_curdoc(doc, invoke)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 408, in invoke_with_curdoc
    return f()
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 449, in invoke
    return f(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/bokeh/document/callbacks.py", line 179, in remove_then_invoke
    return callback()
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 243, in refresh
    raise e
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 239, in refresh
    self._trigger_refresh(stream_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 256, in _trigger_refresh
    self.update(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 949, in update
    item = self.__getitem__(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 435, in __getitem__
    self.update_frame(frame)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/bokeh/element.py", line 2410, in update_frame
    element = self._get_frame(key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/plot.py", line 1259, in _get_frame
    frame = get_plot_frame(self.hmap, key_map, cached)
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/util.py", line 284, in get_plot_frame
    raise e
  File "/user_environment/lib/python3.10/site-packages/holoviews/plotting/util.py", line 280, in get_plot_frame
    return map_obj[key]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 550, in __call__
    if not args and not kwargs and not any(kwarg_hash): return self.callable()
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 199, in dynamic_mul
    self_el = self.select(HoloMap, **key_map) if self.kdims else self[()]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 550, in __call__
    if not args and not kwargs and not any(kwarg_hash): return self.callable()
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 204, in dynamic_mul
    other_el = other.select(HoloMap, **key_map) if other.kdims else other[()]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 581, in __call__
    ret = self.callable(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1012, in dynamic_operation
    key, obj = resolve(key, kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1001, in resolve
    return key, map_obj[key]
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 1216, in __getitem__
    val = self._execute_callback(*tuple_key)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 983, in _execute_callback
    retval = self.callback(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/core/spaces.py", line 581, in __call__
    ret = self.callable(*args, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1013, in dynamic_operation
    return apply(obj, *key, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 1005, in apply
    processed = self._process(element, key, kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/util/__init__.py", line 989, in _process
    return self.p.operation(element, **kwargs)
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 612, in _build_layer_callback
    selection = self._select(element, exprs[layer_number], cache)
  File "/user_environment/lib/python3.10/site-packages/holoviews/selection.py", line 519, in _select
    raise CallbackError("linked_selection aborted because it could not "
holoviews.core.options.CallbackError: linked_selection aborted because it could not display selection for all elements: only integer scalar arrays can be converted to a scalar index.
```

### PR DESCRIPTION
**Pull request by mmorys**
Fixes #5567.

Dask is not compatible with direct assignment when using an array of indices. This change catches the `TypeError` raised when direct assignment is attempted on a Dask Series, and reverts to using the mask() method of `dask.Series`.

### CODE REVIEW

**Comment by mmorys** [comment]
This fix enables lasso select on plots (datashaded and not datashaded). However, a secondary issue was discovered following this fix. Cross-filtering (i.e. using lasso-select on multiple linked plots) raises an error in certain selection situations. Will investigate further and open relevant issue...

**Comment by philippjfr** [comment]
Thanks @mmorys! This is great and the fix looks good.

**Comment by mmorys** [comment]
> This fix enables lasso select on plots (datashaded and not datashaded). However, a secondary issue was discovered following this fix. Cross-filtering (i.e. using lasso-select on multiple linked plots) raises an error in certain selection situations. Will investigate further and open relevant issue...

The source of the issue led me back to the [mask](https://docs.dask.org/en/stable/generated/dask.dataframe.Series.mask.html) method I am using on the series. Even though the original dask series and pandas series with masking values both have `dtype=bool`, the resulting dask series has `dtype=object`. This led to issues downstream.

I just pushed a simple fix to this, saving the original ._meta attribute of the dask series, and setting it back to this stored value after the `mask()` operation.

*Sidenote: it's an unfortunate coincidence that mask refers to both the object and the objcet's method here.*

**Comment by mmorys** [comment]
sorry. I goofed. Pushed to the wrong branch while experimenting. Reverted...

File: holoviews/element/selection.py [hunk 1]
```diff
@@ -117,15 +118,28 @@ def spatial_select_columnar(xvals, yvals, geometry):
...
 def spatial_select(xvals, yvals, geometry):
     if xvals.ndim > 1:
         return spatial_select_gridded(xvals, yvals, geometry)
     else:
-        return spatial_select_columnar(xvals, yvals, geometry)
+        df = xvals.to_frame(name="xvals")
+        df["yvals"] = yvals
+        def spatial_select_df(df, geometry):
+            return spatial_select_columnar(df.xvals, df.yvals, geometry)
+        return df.map_partitions(spatial_select_df, geometry)
```
[the remark this answers is not in the document]
Comment (mmorys) [reply holoviews/element/selection.py]: tldr: I removed these lines.

Seems I had doubled down on my previous accidental push and revered the wrong commit. Was experimenting on a higher-performance implementation for `dask`.
Reverted both changes and ensured I got it right with a diff. Commit [9956ab1](https://github.com/holoviz/holoviews/pull/5568/commits/9956ab1d8bf5b7bd4fb740acf7a47df46f3b6053) is now back at the state of [19b9873](https://github.com/holoviz/holoviews/pull/5568/commits/19b987316548e1f3bac0e9e196ad81b7fc6bead1).

File: holoviews/element/selection.py [hunk 2]
```diff
@@ -117,15 +118,28 @@ def spatial_select_columnar(xvals, yvals, geometry):
...
-    mask[np.where(mask)[0]] = geom_mask
+    try:
+        mask[np.where(mask)[0]] = geom_mask
+    except TypeError:
+        # Dask not compatible with above assignment statement.
+        # To circumvent, create a Series, fill in geom_mask values,
+        # and use mask() method to fill in values.
+        geom_mask_expanded = pd.Series(False, index=mask.index)
+        geom_mask_expanded[np.where(mask)[0]] = geom_mask
+        mask = mask.mask(mask, geom_mask_expanded)
```
[the remark this answers is not in the document]
Comment (mmorys) [reply holoviews/element/selection.py:132]: > I have left some comments.
> 
> Is it possible for you to add some small unit tests, which test this functionality both for pandas and dask DataFrames?

Yes, will do.

**Comment by mmorys** [comment]
> I have left some comments.
> 
> Is it possible for you to add some small unit tests, which test this functionality both for pandas and dask DataFrames?

Yes, will do.

**Comment by mmorys** [comment]
This should be a nice stable point for the pull request. Following the initial proposal, I noticed a glaring performance issue with the original, namely that the point-in-polygon detection of `spatialpandas`/`shapely` was not exploiting `dask` parallelism. [c5f2f03](https://github.com/holoviz/holoviews/pull/5568/commits/c5f2f0342d81a042e7b2e962903ecb3c3307a82b) kills 2 birds with one stone by:
1. Fixing the original bug, ensuring the that `sel_mask[np.where(sel_mask)[0]] = geom_mask` is not called on dask Series objects.
2. Using dask's `map_partitions()` to recursively call `spatial_select_columnar()` using the underlying `pandas` dataframes in parallel. Evaluating on a 100M point plot showed several times speed improvement over the former solution.

File: holoviews/tests/element/test_selection.py [hunk 3]
```diff
@@ -592,3 +601,54 @@ def test_poly_geom_selection_inverted(self):
...
+            })
+
+    @pytest.fixture(scope="class")
+    def enclosed_pt_mask(self):
+        enclosed_pt_list = [
+            0, 0, 0,
+            1, 1, 0,
+            1, 1, 0,
+        ]
+        return np.array(enclosed_pt_list, dtype=np.bool)
```
Comment (Hoxbro) [inline holoviews/tests/element/test_selection.py:635]:
```suggestion
        return np.array(enclosed_pt_list, dtype=bool)
```

Comment (mmorys) [reply holoviews/tests/element/test_selection.py]: fixed in [0bf0eec](https://github.com/holoviz/holoviews/pull/5568/commits/0bf0eecbecb106b5678c7ad40a83e9467dd81f38)

**Comment by Hoxbro** [comment]
Thank you for your contribution! :1st_place_medal:

**Pull request merged**